Add Grids

dtedder opened the pull request:
added grid button to map navigation tool area. linked grid mode to coordinate format

#421

dtedder committed:
added grid button to map navigation tool area. linked grid mode to coordinate format

JamesMBallard commented on Shared/GridController.cpp:
```diff
@@ -0,0 +1,126 @@
+{
+  // create a string lookup to use lat long for everything but utm/mgrs/usng
+  using namespace Esri::ArcGISRuntime::Toolkit;
+  const static std::vector<QString> nonDefaultFormatNames{
```
Subjective, but we typically go with `static const`.

dtedder replied:
👍

dtedder committed:
no need for this to be public

JamesMBallard commented on Shared/GridController.cpp:
```diff
@@ -0,0 +1,126 @@
+  // access tool properties from the config
+  auto formatLookup = QString{QStringLiteral("default")};
+  auto coordinateFormat = properties["CoordinateFormat"].toString();
+  if (std::find(nonDefaultFormatNames.cbegin(), nonDefaultFormatNames.cend(), coordinateFormat) != nonDefaultFormatNames.cend())
```
For three entries it hardly matters, but I would still recommend an unordered_set for quick lookups vs linear search through a vector.

Also there has been an emphasis on using the free functions std::begin, std::cbegin, etc over the members. I go back and forth but I tend to stick with the free functions.

dtedder replied:
switched to free functions and unordered_set

JamesMBallard commented on Shared/GridController.h:
```diff
@@ -0,0 +1,59 @@
+
+// Qt headers
+#include <QString>
+#include <QVariantMap>
```
You can probably forward declare this one. QtContainerFwd should have all the forwards declares to make that easy. No need to copy paste out of that file, just include it here. Then you can include `<QVariantMap>` over in the .cpp file.

dtedder replied:
👍

JamesMBallard commented on Shared/NavigationController.h:
```diff
@@ -50,6 +50,7 @@ class NavigationController : public AbstractTool
   Q_INVOKABLE void pan();
   Q_INVOKABLE void setRotation();
   Q_INVOKABLE void set2D();
+  Q_INVOKABLE void setShowGrid(bool show);
```
This doesn't make much difference, but I would think `showGrid` would make sense as a property in the `Q_PROPERTY`. Invokable methods are typically more meaningful operations and properties are more like simple state management stuff, and this feels more like simple state management.

I recommend making this a property. Maybe:
```c++
Q_PROPERTY(bool showGrid READ showGrid WRITE setShowGrid NOTIFY showGridChanged)
```

Then `void showGridChanged(bool show);` wouldn't need the value passed on the signal. That should be simple enough to change elsewhere where it's used.

If you want to address these separately, that works as well. For example if this makes sense as a property, but you need to keep passing the value on the signal.

dtedder replied:
Moved this to use Q_PROPERTY and moving everything over to GridController directly made things much simpler.

JamesMBallard commented on Shared/qml/NavigationTool.qml:
```diff
@@ -32,6 +32,10 @@ Item {
     width: controlsColumn.width + 10 * scaleFactor
     property alias radius: backgroundRecatangle.radius
 
+    GridController {
+        id: gridController
+    }
```
If the `showGrid` logic is a property, you can bind it here as long as that button has an id.
like
```qml
    GridController {
        id: gridController
        showGrid: overlayButton.selected
    }
```

That's another benefit over invokable methods.

dtedder replied:
👍

dtedder committed:
switched to Q_PROPERTY for showGrid
using std::cbegin instead of member methods
use std::unordered_set
forward declare with QtContainerFwd

dtedder committed:
apply labeling format for degrees minutes seconds

dtedder committed:
switched to storing properties for color and visibility

dtedder committed:
Isolate setProperties to only those associated with a tool (#461)

* Add setProperty to AbstractTool
Modify DsaController to call setProperty in onPropertyChanged signal
Update tool controllers to implement setProperty and subsequently call their existing setProperties method if a relevant property was altered

* switched to bool method and renamed setProperties to be more descriptive of initialization process

* using std::any_if for longer lists of strings to find

* this was mistakenly marked as private during testing. moving back to public section

* this line also used for previous test and should ultimately not be included here.

* missing space

JamesMBallard commented on Shared/AbstractTool.cpp:
```diff
@@ -99,6 +112,11 @@ bool AbstractTool::isActive() const
 
+bool AbstractTool::setContainsString(const std::unordered_set<QString>& strSet, const QString& str)
+{
+  return std::any_of(std::cbegin(strSet), std::end(strSet), [str](const QString& p) { return str == p; });
```
```suggestion
  return std::any_of(std::cbegin(strSet), std::cend(strSet), [str](const QString& p) { return str == p; });
```

JamesMBallard reviewed (approved):
Very minor nitpick, but very good overall. I like the new helper and approach.  No need for another review.

dtedder committed:
missed const on the end iterator.
